### ISSUE TITLE
`INVALID_BUFFER_SIZE` when using the same kernel in different contexts

### ISSUE THREAD

**Issue by genbattle** [pocl/pocl#935]
I'm using POCL as the backend for running tests against some OpenCL code in CI, and I kept running into this error:

```
[2021-03-19 01:43:47.915105283]POCL: in fn POclCreateBuffer at line 98:
  |     ERROR | CL_INVALID_BUFFER_SIZE Size (50803200) is bigger than max mem alloc size (0) of all devices in context
thread 'test::test_resize_nn' panicked at 'called `Result::unwrap()` on an `Err` value: 

################################ OPENCL ERROR ############################### 

Error executing function: clCreateBuffer  

Status error code: CL_INVALID_BUFFER_SIZE (-61)  

Please visit the following url for more information: 

https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clCreateBuffer.html#errors  

############################################################################# 
```

Here's my full test output:

```
running 3 tests
thread 'test::test_resize_linear' panicked at 'called `Option::unwrap()` on a `None` value', resize/src/lib.rs:276:34
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2021-03-19 01:43:40.591812128]POCL: in fn pocl_init_devices at line 489:
  |   GENERAL |  Installing SIGFPE handler...
[2021-03-19 01:43:40.591868601]POCL: in fn pocl_basic_init at line 403:
  |   WARNING |  INIT dlcache DOTO delete
test test::test_resize_linear ... FAILED
Got CPU device
Got context
[2021-03-19 01:43:40.592236689]POCL: in fn POclCreateCommandQueue at line 41:
  |   GENERAL |  Create Command queue on device 0
Created queue
[2021-03-19 01:43:40.592508075]POCL: in fn compile_and_link_program at line 634:
  |   GENERAL |  building from sources for device 0
Got CPU device
Got context
[2021-03-19 01:43:40.596147197]POCL: in fn POclCreateCommandQueue at line 41:
  |   GENERAL |  Create Command queue on device 0
Created queue
[2021-03-19 01:43:40.596281567]POCL: in fn compile_and_link_program at line 634:
  |   GENERAL |  building from sources for device 0
[2021-03-19 01:43:40.632864884]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.632960472]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633021411]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633079962]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633137998]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633194415]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633222377]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   1 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c0053e0 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633244851]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   2 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c000c70 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633264093]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   4 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c0061b0 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633282840]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   5 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c02d900 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633384100]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633445838]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633549350]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633639280]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633703642]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633759549]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.633845775]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   1 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c0061b0 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633871065]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   2 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c000c70 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633889269]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   4 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c0053e0 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.633907568]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   5 ||     uint || Local 0 || Size      4 || Value 0x7f0d0c000fc0 || *Value (nil) || *(uint32*)Value:        0 
Compiled kernels
basic-Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
CPU
[2021-03-19 01:43:40.684955983]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685008374]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685047463]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685083199]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685117931]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685152954]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685166124]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   1 ||     uint || Local 0 || Size      4 || Value 0x7f0d14008490 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685173189]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   2 ||     uint || Local 0 || Size      4 || Value 0x7f0d14001690 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685225941]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   4 ||     uint || Local 0 || Size      4 || Value 0x7f0d14006bb0 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685242204]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel  resize_rgb8_nn || SetArg idx   5 ||     uint || Local 0 || Size      4 || Value 0x7f0d1408ad70 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685327538]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685381430]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685432018]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685481719]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685531214]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685580354]POCL: in fn POclGetKernelArgInfo at line 46:
  |     ERROR | CL_KERNEL_ARG_INFO_NOT_AVAILABLE argument information is not available!
[2021-03-19 01:43:40.685604752]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   1 ||     uint || Local 0 || Size      4 || Value 0x7f0d14001690 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685621771]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   2 ||     uint || Local 0 || Size      4 || Value 0x7f0d14008490 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685640634]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   4 ||     uint || Local 0 || Size      4 || Value 0x7f0d140059f0 || *Value (nil) || *(uint32*)Value:        0 
[2021-03-19 01:43:40.685656147]POCL: in fn POclSetKernelArg at line 69:
  |   GENERAL |  Kernel resize_rgb8_linear || SetArg idx   5 ||     uint || Local 0 || Size      4 || Value 0x7f0d1408aa50 || *Value (nil) || *(uint32*)Value:        0 
Compiled kernels
basic-Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
CPU
test test::test_get_device ... ok
[2021-03-19 01:43:47.915105283]POCL: in fn POclCreateBuffer at line 98:
  |     ERROR | CL_INVALID_BUFFER_SIZE Size (50803200) is bigger than max mem alloc size (0) of all devices in context
thread 'test::test_resize_nn' panicked at 'called `Result::unwrap()` on an `Err` value: 

################################ OPENCL ERROR ############################### 

Error executing function: clCreateBuffer  

Status error code: CL_INVALID_BUFFER_SIZE (-61)  

Please visit the following url for more information: 

https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clCreateBuffer.html#errors  

############################################################################# 
', resize/src/lib.rs:261:70
test test::test_resize_nn ... FAILED

failures:

failures:
    test::test_resize_linear
    test::test_resize_nn

test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 7.34s
```

This only occurred when more than one thread was used for running the tests (this is running under Rust's standard test harness) I would get the error above. When I ran the tests in a single thread the tests would pass without any issues.

When I investigated whether there were any thread/race conditions in the OpenCL API everything I read pointed to the only race condition being in setting kernel arguments across multiple threads with the same kernel. What I found was that POCL seems to fall over when the same kernel is referenced from multiple threads simultaneously, even if they have completely separate contexts.

This issue doesn't seem to occur with other ICDs, but am I doing something wrong here with having separate instances of the same kernel in different threads, or is there something wrong with the kernel cache in POCL that is causing it to use the same kernel object across multiple contexts in different threads simultaneously?

Edit: `POCL_KERNEL_CACHE=0` didn't fix the problem, it's still just intermittent.

Edit2: Ok so reading the error more closely it seems it isn't related to the kernel, but a buffer allocation that's failing due to there being 0 memory available.

Edit3: Platform is Ubuntu 20.04, POCL version is 1.4-6.

**Comment by genbattle** [comment]
I had a look into the code, and I suspect this might be some sort of race condition with device initialization, specifically the initialization of `max_mem_alloc_size`. It should be set to a non-zero size once the device is initialized, so I must be hitting one heck of an edge case for this to happen. I'll try and test with the latest POCL tomorrow and see if that resolves anything.

**Comment by genbattle** [comment]
I upgraded to POCL 1.6 by manually installing it, and the issue seems to be resolved, feel free to close this issue if 1.4 is out of support now. I just need to figure out how to upgrade all of my Github Actions runners with Ubuntu 20.04 to POCL 1.6 😅

**Comment by pjaaskel** [comment]
Glad to hear it has been fixed. Unfortunately we do not have anyone actively maintaining the old releases.
